File: vm/src/Execution/Ops/FlowControl.hs
{-
-- EPITECH PROJECT, 2025
-- Delivery
-- File description:
-- Flow-control
-}

module Execution.Ops.FlowControl where

import Types (Value(..))
import Execution.State (VMState(..))

opJump :: Int -> VMState -> Either String VMState
opJump addr state = Right $ state { ip = addr - 1 }

opJumpIfTrue :: Int -> VMState -> Either String VMState
opJumpIfTrue addr state =
    case stack state of
        (BoolVal True : rest) -> Right $ state { stack = rest, ip = addr - 1 }
        (BoolVal False : rest) -> Right $ state { stack = rest }
        (_ : _) -> Left "Error: JumpIfTrue requires a boolean on the stack"
        [] -> Left "Error: JumpIfTrue requires a value on the stack"

opJumpIfFalse :: Int -> VMState -> Either String VMState
opJumpIfFalse addr state =
    case stack state of
        (BoolVal False : rest) -> Right $ state { stack = rest, ip = addr - 1 }
        (BoolVal True : rest) -> Right $ state { stack = rest }
        (_ : _) -> Left "Error: JumpIfFalse requires a boolean on the stack"
        [] -> Left "Error: JumpIfFalse requires a value on the stack"

File: bonus/vm/src/execution/ops/flow_control.py
from typing import Union
from vm_types import Value, BoolVal
from execution.state import VMState

def op_jump(addr: int, state: VMState) -> Union[VMState, str]:
    state.ip = addr - 1
    return state

def op_jump_if_true(addr: int, state: VMState) -> Union[VMState, str]:
    if not state.stack:
        return "Error: JumpIfTrue requires a value on the stack"
    
    val = state.stack.pop()
    if isinstance(val, BoolVal):
        if val.val:
            state.ip = addr - 1
    else:
        return "Error: JumpIfTrue requires a boolean on the stack"
        
    return state

def op_jump_if_false(addr: int, state: VMState) -> Union[VMState, str]:
    if not state.stack:
        return "Error: JumpIfFalse requires a value on the stack"
    
    val = state.stack.pop()
    
    if isinstance(val, BoolVal):
        if not val.val:
            state.ip = addr - 1
    else:
        return "Error: JumpIfFalse requires a boolean on the stack"
        
    return state
